### PR TITLE
Add The First Decade metadata to the new content installer.

### LINKED ISSUE #11523: Add content installer support for TFD.

### PR DESCRIPTION
**Pull request by pchote**
Closes #11523.

### CODE REVIEW

File: OpenRA.Mods.Common/UtilityCommands/ListInstallShieldCabContentsCommand.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	class ListInstallShieldCabContentsCommand : IUtilityCommand
+	{
+		public string Name { get { return "--list-installshield-cab"; } }
+
+		public bool ValidateArguments(string[] args)
+		{
+			return args.Length == 2;
+		}
+
+		[Desc("ARCHIVE.CAB", "Lists the filenames contained within a MSCAB file")]
```
Comment (Mailaender) [inline OpenRA.Mods.Common/UtilityCommands/ListInstallShieldCabContentsCommand.cs]:
```patch
-MSCAB 
+InstallShield CAB
```

**Comment by pchote** [comment]
I found that I needed to apply 5be513627c133f213586c39fb142fcccc3672610 for my TFD image to be visible to mono.

File: OpenRA.Mods.Common/FileFormats/InstallShieldCABCompression.cs [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+			}
+
+			if (file.Flags.HasFlag(CABFlags.FileObfuscated))
+				throw new NotImplementedException("Obfuscated files are not supported");
+
+			var extracter = new CabExtracter(file, volumes);
+			extracter.CopyTo(output, onProgress);
+
+			if (output.Length != file.ExpandedSize)
+				throw new Exception("Did not fully extract Expected = {0}, Got = {1}".F(file.ExpandedSize, output.Length));
```
Comment (Mailaender) [inline OpenRA.Mods.Common/FileFormats/InstallShieldCABCompression.cs]: `InvalidDataException` maybe? and `Did not fully extract.` with a sentence end might make this easier to read.

Comment (Phrohdoh) [inline OpenRA.Mods.Common/FileFormats/InstallShieldCABCompression.cs]: Us failing to extract doesn't necessarily mean the data is invalid.

Comment (pchote) [inline OpenRA.Mods.Common/FileFormats/InstallShieldCABCompression.cs]: This was copied from the original code.  Fixed.

**Comment by Mailaender** [comment]
Works as promised. :white_check_mark:

File: OpenRA.Mods.Common/UtilityCommands/ListInstallShieldCabContentsCommand.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	class ListInstallShieldCabContentsCommand : IUtilityCommand
+	{
+		public string Name { get { return "--list-installshield-cab"; } }
+
+		public bool ValidateArguments(string[] args)
+		{
+			return args.Length == 2;
+		}
+
+		[Desc("ARCHIVE.CAB", "Lists the filenames contained within an Installshield CAB file")]
```
Comment (Mailaender) [inline OpenRA.Mods.Common/UtilityCommands/ListInstallShieldCabContentsCommand.cs]:
```patch
-ARCHIVE.CAB
+DATA.HDR
```

**Comment by pchote** [comment]
Fixed.

File: OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromDiscLogic.cs [hunk 4]
```diff
@@ -370,6 +399,64 @@ static void ExtractFromMSCab(string path, MiniYaml actionYaml, List<string> extr
...
 
+		static void ExtractFromISCab(string path, MiniYaml actionYaml, List<string> extractedFiles, Action<string> updateMessage)
+		{
+			var sourcePath = Path.Combine(path, actionYaml.Value);
+
+			// Try as an absolute path
+			if (!File.Exists(sourcePath))
+				sourcePath = Platform.ResolvePath(actionYaml.Value);
+
+			var volumeNode = actionYaml.Nodes.FirstOrDefault(n => n.Key == "Volumes");
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromDiscLogic.cs]: Shouldn't these be `SingleOrDefault`?

File: OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromDiscLogic.cs [hunk 5]
```diff
@@ -288,6 +301,11 @@ enum ExtractionType { Raw, Blast }
 		static void ExtractFromPackage(ExtractionType type, string path, MiniYaml actionYaml, List<string> extractedFiles, Action<string> updateMessage)
 		{
 			var sourcePath = Path.Combine(path, actionYaml.Value);
+
+			// Try as an absolute path
+			if (!File.Exists(sourcePath))
+				sourcePath = Platform.ResolvePath(actionYaml.Value);
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromDiscLogic.cs]: You're duplicating those four lines a bunch. Make it a method?

**Comment by MortovMolotov** [comment]
I have a working TFD DVD; but I'm a GitHub n00b. How can I help?

**Comment by obrakmann** [comment]
@MortovMolotov: if you're using Windows, the easiest way would be to install https://ci.appveyor.com/api/buildjobs/m7p5a3bih0v43nsj/artifacts/OpenRA-eec3c33.exe (into a separate directory so you don't overwrite your release version), then move the `Contents` directory out of the way in your support directory and then try installing the game data with the new OpenRA version.

**Comment by GraionDilach** [comment]
Needs a rebase as well.

**Comment by pchote** [comment]
> the --list-installshield-cab command doesn't work for random installshield cabs I have here ... Is it supposed to work with those?

Our installshield parser has only ever supported the parts of the format that used by the WW installers.  I'm not surprised that it doesn't work for other games.

**Comment by x-a-n-a-x** [comment]
I have just tested pchote's tfd branch. It works like a charm.   :+1:

**Comment by ABrandau** [comment]
Just tested installing the game assets using the TFD disc and it works well, I managed to open the shellmap of ts, ra1 and cnc, it's nice to see that it includes the music aswell.

**Comment by pchote** [comment]
Rebased.

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/698afc749c87aaa8c6dede0ab552789aa68a9c5e)

**Pull request merged**